File: backend/app/models/booking.py
from datetime import datetime
from enum import Enum
from typing import Optional

from sqlmodel import Field, Relationship, SQLModel


class BookingStatus(str, Enum):
    REQUESTED = "REQUESTED"
    ACCEPTED = "ACCEPTED"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"


class BookingRequestBase(SQLModel):
    scheduled_at: datetime
    duration_hours: float
    location: str
    notes: Optional[str] = None


class BookingRequest(BookingRequestBase, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    listing_id: int = Field(foreign_key="servicelisting.id")
    requester_id: int = Field(foreign_key="user.id")
    provider_id: int = Field(foreign_key="user.id")
    status: BookingStatus = Field(default=BookingStatus.REQUESTED)
    total_price: float
    payment_status: str = Field(default="PENDING")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class BookingRequestCreate(BookingRequestBase):
    listing_id: int
    provider_id: int
    total_price: float


class BookingRequestRead(BookingRequestBase):
    id: int
    listing_id: int
    requester_id: int
    provider_id: int
    status: BookingStatus
    total_price: float
    payment_status: str


class BookingStatusUpdate(SQLModel):
    new_status: BookingStatus

File: client/lib/shared/models/booking.dart
import 'package:intl/intl.dart';

class BookingModel {
  BookingModel({
    required this.id,
    required this.listingId,
    required this.requesterId,
    required this.providerId,
    required this.status,
    required this.scheduledAt,
    required this.durationHours,
    required this.location,
    required this.totalPrice,
    this.paymentStatus,
    this.notes,
  });

  final int id;
  final int listingId;
  final int requesterId;
  final int providerId;
  final String status;
  final DateTime scheduledAt;
  final double durationHours;
  final String location;
  final double totalPrice;
  final String? paymentStatus;
  final String? notes;

  factory BookingModel.fromJson(Map<String, dynamic> json) {
    return BookingModel(
      id: json['id'] as int,
      listingId: json['listing_id'] as int,
      requesterId: json['requester_id'] as int,
      providerId: json['provider_id'] as int,
      status: json['status'] as String,
      scheduledAt: DateTime.parse(json['scheduled_at'] as String),
      durationHours: (json['duration_hours'] as num).toDouble(),
      location: json['location'] as String,
      totalPrice: (json['total_price'] as num).toDouble(),
      paymentStatus: json['payment_status'] as String?,
      notes: json['notes'] as String?,
    );
  }

  String get formattedDate => DateFormat.yMMMd().add_jm().format(scheduledAt);
}

